src/main/java/hudson/remoting/Channel.java

```diff
@@ -1584,17 +1594,48 @@ public static Channel current() {
         return CURRENT.get();
     }
 
+    // TODO: Unrestrict after the merge into the master.
+    // By now one may use it via the reflection logic only
     /**
      * Calls {@link #dumpDiagnostics(PrintWriter)} across all the active channels in this system.
      * Used for diagnostics.
      *
-     * @since 2.26.3
-     */
-    public static void dumpDiagnosticsForAll(PrintWriter w) throws IOException {
-        for (Ref ref : ACTIVE_CHANNELS.values().toArray(new Ref[0])) {
-            Channel ch = ref.channel();
-            if (ch!=null)
-                ch.dumpDiagnostics(w);
+     * @param w Output destination
+     * 
+     * @since 2.62.3 - stable 2.x (restricted)
+     * @since TODO 3.x - public version 
+     */
+    @Restricted(NoExternalUse.class)
+    public static void dumpDiagnosticsForAll(@Nonnull PrintWriter w) {
+        final Ref[] channels = ACTIVE_CHANNELS.values().toArray(new Ref[0]);
+        int processedCount = 0;
+        for (Ref ref : channels) {
+            // Check if we can still write the output
+            if (w.checkError()) {
+                logger.log(Level.WARNING, 
+                        String.format("Cannot dump diagnostics for all channels, because output stream encontered an error. "
+                                + "Processed %d of %d channels, first unprocessed channel reference is %s.",
+                                processedCount, channels.length, ref
+                        )); 
+                break;
+            }
+            
+            // Dump channel info if the reference still points to it
+            final Channel ch = ref.channel();
+            if (ch != null) {
+                try {
+                    ch.dumpDiagnostics(w);
+                } catch (Throwable ex) {
+                    if (ex instanceof Error) {
+                        throw (Error)ex;
```

Comment:
Not necessarily helpful. (For `ThreadDeath`, yes; for `LinkageError`, generally no; for `OutOfMemoryError`, probably irrelevant.)